### ISSUE TITLE
Question URLs are showing database IDs

### ISSUE THREAD

**Issue by konklone** [opengovernment/askthem-legacy#330]
This URL is, I think, showing a MongoDB database ID:

http://www.askthem.io/dc/questions/52f4fc01f92ea1bb98000009

Whatever it is, I'd turn that into a slug before your formal launch, for a few reasons:
- Better clickthrough rates, people will have an idea of what they'll be taken to
- Better SEO
- More future-proof URL architecture - you can change the underlying system more easily

**Comment by davidmooreppf** [comment]
Thx Eric, good suggestions - appreciate it - we're just getting around to some of this, happy to have Walter take a look. Our DB of elected officials is still pretty flux-y. But very on-point.

**Comment by walter** [comment]
There are at least three classes where we may want to do this right now:
- questions
- people (already done for state legislators and federal legislators)
- users

All three are tricky because the natural "stub url" would be non-unique names or titles.

Normally you would deal with this by adding a "scope" (i.e. all question titles must be unique for a jurisdiction), but these three classes using a scope (or scopes, e.g. unique title for person for specific jurisdiction by a particular user) won't work elegantly. We would probably end up with ugly urls anyway.

I'll consider it more. I would say that the priority is having a stub url for questions and people. Users are a nice to have. The solution that comes to mind is having a sequence attribute and add that to the title. Possibly timestamp as alternative to sequence.

@konklone if you have an alternative approach, I'm all ears.

**Comment by konklone** [comment]
If it's helpful, I'm using [mongoid-slug](https://github.com/digitalplaywright/mongoid-slug) to power the URLs for [posts on my blog](https://konklone.com/post/protect-your-domain-name-with-two-factor-authentication). It uses the database to reflect on whether a slug already exists and appends `-1` or `-2` as appropriate. It associates an array of slugs with each object, so you can alter slugs and still have old URLs resolve.

In other words, I would ignore the uniqueness constraint as an application-level concern, and allow dupe question/people/user names to generate nearly identical slugs, that `mongoid-slug` takes care of differentiating.

**Comment by walter** [comment]
Cool. I had found mongo-slug (and was aware of it), but hadn't caught it handles our situation of non-unique titles and the like. Instead I focused on how you can specify scope, etc.

It's actually pretty much what I was suggesting for a sequence attribute. A tad simpler and I'm happy to have it implemented for me!

I'll do some experiments and see where it gets us. Thanks for you suggestions.

**Comment by walter** [comment]
Did a little experiment on Question model to see if this is viable right away.

Unfortunately it looks like it will take some doing. At least it won't simply "just work" for us without further digging.

Too bad!

I do think we want to get this relatively soon, but just not in the next couple days probably.

A goal is that old URLs still work after the new slug URLs happen. This seems possible with mongoid-slug.
